src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2221,114 +2202,100 @@ impl Coordinator {
         }
     }
 
+    /// Determines the query timestamp and acquires read holds on dependent sources
+    /// if necessary.
+    fn sequence_peek_timestamp(
+        &mut self,
+        session: &mut Session,
+        when: &QueryWhen,
+        cluster_id: ClusterId,
+        timeline_context: TimelineContext,
+        id_bundle: &CollectionIdBundle,
+        source_ids: &BTreeSet<GlobalId>,
+        real_time_recency_ts: Option<Timestamp>,
+    ) -> Result<TimestampDetermination<Timestamp>, AdapterError> {
+        let determination = self.determine_timestamp(
+            session,
+            id_bundle,
+            when,
+            cluster_id,
+            timeline_context.clone(),
+            real_time_recency_ts,
+        )?;
+
+        // We only track the peeks in the session if the query doesn't use AS
+        // OF or we're inside an explicit transaction. The latter case is
+        // necessary to support PG's `BEGIN` semantics, whose behavior can
+        // depend on whether or not reads have occurred in the txn.
+        if matches!(session.transaction(), &TransactionStatus::InTransaction(_))
+            || when == &QueryWhen::Immediately
+        {
+            session.add_transaction_ops(TransactionOps::Peeks(
+                determination.timestamp_context.clone(),
+            ))?;
+        }
+
+        let in_immediate_multi_stmt_txn = session.transaction().is_in_multi_statement_transaction()
+            && when == &QueryWhen::Immediately;
```

Comment:
I'm not a huge fan of having to generate this value out here and inside `determine_timestamp`. Is it possible to only generate this once?

Reply:
Yeah, I agree it's not ideal.

It does feel a little weird to pass it as a parameter when it can be computed trivially from existing references, though. Personally I think it's fine to leave it to each method to figure out if we're in a transaction and attempting to perform the query immediately.

But happy to change if you feel strongly otherwise!